### ISSUE TITLE
OutOfMemoryException when reading all item of a collection

### ISSUE THREAD

**Issue by xamoulin** [markmeeus/MarcelloDB#69]
I have this exception when I call All twice on an empty collection. Also, it does not happen on "new" session i.e. if the db file does not exist yet. I tried to isolate the problem but it does not occur on my test project, it might have something related to the type I serialized.

Here is the call stack:

```
android.runtime.JavaProxyThrowable: System.OutOfMemoryException: Out of memory
  at (wrapper alloc) System.Object.AllocVector(intptr,intptr)
  at MarcelloDB.netfx.Storage.FileStorageStream.Read (System.Int64 address, System.Int32 length) [0x00001] in <621c1737663941ddb44ffcdd8933ca7b>:0 
  at MarcelloDB.Storage.StreamActors.Reader.Read (System.Int64 address, System.Int32 length) [0x00001] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Storage.StreamActors.JournalledReader.Read (System.Int64 address, System.Int32 length) [0x00001] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Storage.StorageEngine.Read (System.Int64 address, System.Int32 length) [0x00001] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Records.RecordManager.ReadEntireRecord (System.Int64 address) [0x0001b] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Records.RecordManager.GetRecord (System.Int64 address) [0x00001] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Index.BTree.RecordBTreeDataProvider`1[TNodeKey].GetNode (System.Int64 address) [0x00021] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Index.BTree.RecordBTreeDataProvider`1[TNodeKey].GetRootNode (System.Int32 degree) [0x00040] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Index.BTree.BTreeWalker`1[TK].Reset () [0x00022] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Index.BTree.BTreeWalker`1[TK]..ctor (System.Int32 degree, MarcelloDB.Index.BTree.IBTreeDataProvider`1[TK] dataProvider) [0x00024] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Index.RecordIndex`1[TNodeKey].GetWalker () [0x00008] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.IndexEntryEnumerator`2+<System-Collections-Generic-IEnumerable<MarcelloDB-Index-Entry<TKey>>-GetEnumerator>d__20[T,TKey].MoveNext () [0x000b7] in <5ea921fcec2c4886bbdc14f768df096f>:0 
  at MarcelloDB.Collections.CollectionEnumerator`2+<System-Collections-Generic-IEnumerable<T>-GetEnumerator>d__33[T,TKey].MoveNext () [0x0023b] in <5ea921fcec2c4886bbdc14f768df096f>:0 
```

**Comment by markmeeus** [comment]
Hi,  

Thanks for reporting this.
On what version of MarcelloDB are you experiencing this?

**Comment by xamoulin** [comment]
Hi, 

1.0.5 and 1.0.6. I did not tried on previous versions.

**Comment by xamoulin** [comment]
Also, I forgot to say, I've got this bug only with Android. It seems to works fine on iOS.

**Comment by markmeeus** [comment]
Hi,

That's pretty weird, I asked about the version since I fixed an issue in 1.0.6 which may have caused a similar exception.
So if you are indeed experiencing this issue on 1.0.6, I'll have to take a closer look at the issue.

Would it be possible to send a copy (mark.meeus@gmail.com) of the collection- and journal-file of a situation that has this issue? 
Better not attach it here since even though it's empty,  data may still be in the file.

**Comment by markmeeus** [comment]
Hi there,

Upon inspection of the data files you sent me, for which I thank you, I can confirm some sort of data corruption.
This may have happened by a bug in 1.0.5, which causes the file to be corrupt, 1.0.6 won't be able to correct this.

Have you seen this issue with files that have only been touched by 1.0.6 ?

**Comment by xamoulin** [comment]
Hi,

Yes, I got this error with only 1.0.6. In fact, the first time I got this error is soon after I update to 1.0.6 and I tough it was a new bug so downgrade to 1.0.5, but I had it with 1.0.5 too. So it's just a coincidence that at the same time I was working with empty collection. 

Do you want them with 1.0.6 only ? I would have sent that in the first place but laziness 😄

**Comment by markmeeus** [comment]
Hi,

Yes please :-)
Af file created with 1.0.6 that has this issue would be perfect!

**Comment by markmeeus** [comment]
Upon further inspection it actually looks like it's a variant of this one: #67 
Internally, MarcelloDB keeps long index of the end of the file. For some reason, there seems to be data after this so called 'head' value.
The first iteration actually writes some data at this 'head' location, overwriting data which belongs to another record.
Weird, since this one was supposed to be fixed in 1.0.5.

I'll have to do some digging, but if you have any additional context, that would be welcome.

**Comment by xamoulin** [comment]
I just sent you the files generated with 1.0.6. 

If that may help, objects without explicit type are stored in the DB in the ApplicationSettings collection, it's a combination of a string key and an object. 

And here's the complete workflow:
- Create ApplicationSettings collection
- Read inexistant ApplicationSetting that contains some custom object
- Read another inexistant ApplicationSetting that contains a bool than write it
- Create collection A (explicitly typed)
- Create collection B (explicitly typed), read by id and add an object
- Read all collection A
- Read all collection B
- Close and restart the app
- Read inexistant ApplicationSetting that contains some custom object
- Read an existant ApplicationSetting that contains a bool
- Read by id in collection B
- Read all collection A
- Read all collection B
- Read all collection A -> Crash

Hope it helps!

**Comment by markmeeus** [comment]
Wow, thanks for the info.
Those collections are in a single 
collectionfile? 
Are there other collection files used in the same session?

**Comment by markmeeus** [comment]
Aha! I think I know what's going on now. 
I'll take a closer look tomorrow ...

**Comment by xamoulin** [comment]
Yes, it's all in the same file, is it bad? And no, it's the only file.

**Comment by markmeeus** [comment]
No, not bad at all, I was thinking about a different scenario.
I understand the cause of the bug, still have to figure out how to fix it :)
I’ll keep you posted.

**Comment by markmeeus** [comment]
Hi there,

This commit should fix that problem.
I'll release a 1.0.7 containing this soon.

**Comment by markmeeus** [comment]
Hi there again,

This fix (version 1.0.7) is now on nuget.
Let me know if it fixes the issue for you as well.

Thanks again for reporting it!

**Comment by markmeeus** [comment]
Hi There,

Did 1.0.7 fix your problem?

**Comment by xamoulin** [comment]
Hello,

I don't have this bug anymore but I haven't been able to test for sure that 1.0.7 fixed it because I have not been able to reproduce it after with 1.0.6. I don't know for sure what happened there but I think it was related to the app workspace being rolled back when I deployed on an android. So I may even not had this bug with version 1.0.6, but with 1.0.5. 
Bottom line, now it works with 1.0.7 but I'm still confused about what happened!

Anyway, thanks a lot!

**Comment by markmeeus** [comment]
Hi,

This error only occurred in a very specific scenario. You had to enumerate an empty collection (maybe even 2) without any other transaction and then close the session. (Or close the app).
If the session is then opened again, and other transactions occur, the empty collections will be corrupted.
This has been fixed in 1.0.7
I guess I can close this issue then?

**Comment by xamoulin** [comment]
Yes, you can! And thanks for everything!
